File: ios/Sources/RoadSenseNSBootstrap/Persistence/RetentionPolicy.swift
import Foundation

public enum RetentionPolicy {
    public static func pruneUploadedReadings(
        _ readings: [QueueReadingRecord],
        now: Date
    ) -> [QueueReadingRecord] {
        let cutoff = now.addingTimeInterval(-30 * 24 * 60 * 60)

        return readings.filter { reading in
            guard let uploadedAt = reading.uploadedAt else {
                return true
            }

            return uploadedAt >= cutoff
        }
    }
}

File: android/core-api/src/main/kotlin/ca/roadsense/ns/api/RetentionPolicy.kt
package ca.roadsense.ns.api

import java.time.Duration
import java.time.Instant

data class QueueReadingRecord(
    val id: String,
    val recordedAt: Instant,
    val uploadedAt: Instant?,
)

object RetentionPolicy {
    private val LOCAL_RETENTION = Duration.ofDays(30)

    fun pruneUploadedReadings(
        readings: List<QueueReadingRecord>,
        now: Instant,
    ): List<QueueReadingRecord> {
        val cutoff = now.minus(LOCAL_RETENTION)
        return readings.filter { reading ->
            val uploaded = reading.uploadedAt ?: return@filter true
            !uploaded.isBefore(cutoff)
        }
    }
}
